tiledb/sm/dimension_label/dimension_label.cc

```diff
@@ -118,37 +120,77 @@ void DimensionLabel::open(
     EncryptionType encryption_type,
     const void* encryption_key,
     uint32_t key_length) {
+  // Open indexed array and get schema
   throw_if_not_ok(indexed_array_->open(
       query_type,
       timestamp_start,
       timestamp_end,
       encryption_type,
       encryption_key,
       key_length));
+  auto&& [index_status, indexed_array_schema] =
+      indexed_array_->get_array_schema();
+  throw_if_not_ok(index_status);
+  // Open labelled array and get schema
   throw_if_not_ok(labelled_array_->open(
       query_type,
       timestamp_start,
       timestamp_end,
       encryption_type,
       encryption_key,
       key_length));
-  load_schema();
+  auto&& [label_status, labelled_array_schema] =
+      labelled_array_->get_array_schema();
+  throw_if_not_ok(label_status);
+  load_schema(indexed_array_schema.value(), labelled_array_schema.value());
   query_type_ = query_type;
+  // Restrict reading to only shared fragments.
+  if (query_type_ == QueryType::READ) {
+    if (is_remote_) {
+      throw StatusException(Status_DimensionLabelError(
+          "Reading from remote dimension labels is not yet supported."));
+    }
+    // Get label names
+    std::vector<TimestampedURI> indexed_frag_uris{
+        indexed_array_->array_directory()
+            .filtered_fragment_uris(false)
+            .fragment_uris()};
+    std::vector<TimestampedURI> labelled_frag_uris{
+        labelled_array_->array_directory()
+            .filtered_fragment_uris(false)
+            .fragment_uris()};
+    // Remove mismatched fragments.
+    utils::parse::prune_fragments(indexed_frag_uris, labelled_frag_uris);
+    utils::parse::prune_fragments(labelled_frag_uris, indexed_frag_uris);
+    // Reload with only shared fragments.
+    indexed_array_->load_fragments(indexed_frag_uris);
+    labelled_array_->load_fragments(labelled_frag_uris);
+  }
 }
 
 void DimensionLabel::open_without_fragments(
```

Comment:
As we already discussed a little bit, I would like to make sure we have a solid story for these open without fragments functions. I'm not sure I fully understand what this will be used for in dimension labels but I'll give you the context and you can see if this could be a problem. 

For consolidation, we use open without fragments to first get the array opened, then we will select fragments to consolidate and load them using the FragmentInfo apis.

The issue is that open without fragments calls into ArrayDirectory using the SCHEMA_ONLY mode, which doesn't do listings for delete tiles for example, just array schemas. Then the fragment info API will do all the listings, including array schema (which is duplicated) and will also list the delete tiles in the commits folder. To make sure that the opened array gets the delete tiles (as they are required for consolidation) I made a hack for now to transfer the delete tiles from the fragment info to the opened array but I will remove that very soon. Instead, I want open without fragments to do all the listings, then add a new fragment info internal API that will take that opened array as a parameter and use the already existing array directory and it's listings. That way it will be very easy to remove duplication in the listings and ensure that the opened array has everything it needs (including delete tiles).

Does that make sense?

Reply:
Makes sense. As discussed, I'm not actually using this method yet, so there is no issue with removing it until we actually have the need for it.